batch/batch/driver/canceller.py

```diff
@@ -0,0 +1,198 @@
+import logging
+
+from hailtop.utils import (WaitableSharedPool, retry_long_running, run_if_changed,
+                           AsyncWorkerPool)
+from hailtop import aiotools
+
+from ..batch import unschedule_job, mark_job_complete
+from ..utils import Box
+
+log = logging.getLogger('canceller')
+
+
+class Canceller:
+    def __init__(self, app):
+        self.app = app
+        self.cancel_ready_state_changed = app['cancel_ready_state_changed']
+        self.cancel_running_state_changed = app['cancel_running_state_changed']
+        self.db = app['db']
+        self.async_worker_pool = AsyncWorkerPool(parallelism=100, queue_size=100)
+        self.task_manager = aiotools.BackgroundTaskManager()
+
+    async def async_init(self):
+        self.task_manager.ensure_future(retry_long_running(
+            'cancel_cancelled_ready_jobs_loop',
+            run_if_changed, self.cancel_ready_state_changed, self.cancel_cancelled_ready_jobs_loop_body))
+        self.task_manager.ensure_future(retry_long_running(
+            'cancel_cancelled_running_jobs_loop',
+            run_if_changed, self.cancel_running_state_changed, self.cancel_cancelled_running_jobs_loop_body))
+
+    def shutdown(self):
+        try:
+            self.task_manager.shutdown()
+        finally:
+            self.async_worker_pool.shutdown()
+
+    async def cancel_cancelled_ready_jobs_loop_body(self):
+        records = self.db.select_and_fetchall(
+            '''
+SELECT user, n_cancelled_ready_jobs
+FROM (SELECT user,
+    CAST(COALESCE(SUM(n_cancelled_ready_jobs), 0) AS SIGNED) AS n_cancelled_ready_jobs
+  FROM user_inst_coll_resources
+  GROUP BY user) AS t
+WHERE n_cancelled_ready_jobs > 0;
+''',
+            timer_description='in cancel_cancelled_ready_jobs: aggregate n_cancelled_ready_jobs')
+        user_n_cancelled_ready_jobs = {
+            record['user']: record['n_cancelled_ready_jobs'] async for record in records
+        }
+
+        total = sum(user_n_cancelled_ready_jobs.values())
+        if not total:
+            should_wait = True
+            return should_wait
+        user_share = {
+            user: max(int(300 * user_n_jobs / total + 0.5), 20)
+            for user, user_n_jobs in user_n_cancelled_ready_jobs.items()
+        }
+
+        async def user_cancelled_ready_jobs(user, remaining):
+            async for batch in self.db.select_and_fetchall(
+                    '''
+SELECT id, cancelled
+FROM batches
+WHERE user = %s AND `state` = 'running';
+''',
+                    (user,),
+                    timer_description=f'in cancel_cancelled_ready_jobs: get {user} running batches'):
+                if batch['cancelled']:
+                    async for record in self.db.select_and_fetchall(
+                            '''
+SELECT jobs.job_id
+FROM jobs FORCE INDEX(jobs_batch_id_state_always_run_cancelled)
+WHERE batch_id = %s AND state = 'Ready' AND always_run = 0
+LIMIT %s;
```

Comment:
Cotton had to add this last year because the query wasn't using the correct index. I'm not sure what would have changed that this is no longer the case.